packages/flutter/lib/src/widgets/scrollable.dart

```diff
@@ -1605,3 +1613,668 @@ class _RestorableScrollOffset extends RestorableValue<double?> {
+  /// ```
+  ///
+  /// Calling this method will create a dependency on the closest
+  /// [TwoDimensionalScrollable] in the [context]. The internal [Scrollable]s
+  /// can be accessed through [TwoDimensionalScrollableState.verticalScrollable]
+  /// and [TwoDimensionalScrollableState.horizontalScrollable].
+  ///
+  /// If no [TwoDimensionalScrollable] ancestor is found, then this method will
+  /// assert in debug mode, and throw an exception in release mode.
+  ///
+  /// Alternatively, [Scrollable.of] can be used by providing the desired [Axis]
+  /// to the `axis` parameter.
+  ///
+  /// See also:
+  ///
+  /// * [TwoDimensionalScrollable.maybeOf], which is similar to this method,
+  ///   but returns null if no [TwoDimensionalScrollable] ancestor is found.
+  static TwoDimensionalScrollableState of(BuildContext context) {
+    final TwoDimensionalScrollableState? scrollableState = maybeOf(context);
+    assert(() {
+      if (scrollableState == null) {
+        throw FlutterError(
+          'TwoDimensionalScrollable.of() was called with a context that does not '
+          'contain a TwoDimensionalScrollable widget.\n'
+          'No TwoDimensionalScrollable widget ancestor could be found starting '
+          'from the context that was passed to TwoDimensionalScrollable.of(). '
+          'This can happen because you are using a widget that looks for a '
+          'TwoDimensionalScrollable ancestor, but no such ancestor exists.\n'
+          'The context used was:\n'
+          '  $context',
+        );
+      }
+      return true;
+    }());
+    return scrollableState!;
+  }
+
+  // Informs the scrollable widgets for gesture recognizers.
+  static DiagonalDragBehavior? _diagonalDragBehavior(BuildContext context) {
+    final _TwoDimensionalScrollableScope? widget = context.dependOnInheritedWidgetOfExactType<_TwoDimensionalScrollableScope>();
+    return widget?.diagonalDragBehavior;
+  }
+}
+
+/// State object for a [TwoDimensionalScrollable] widget.
+///
+/// To manipulate one of the internal [Scrollable] widget's scroll position, use
+/// the object obtained from the [verticalScrollable] or [horizontalScrollable]
+/// property.
+///
+/// To be informed of when a [TwoDimensionalScrollable] widget is scrolling,
+/// use a [NotificationListener] to listen for [ScrollNotification]s.
+/// Both axes will have the same viewport depth since there is only one
+/// viewport, and so should be differentiated by the [Axis] of the
+/// [ScrollMetrics] provided by the notification.
+class TwoDimensionalScrollableState extends State<TwoDimensionalScrollable> {
+  late ScrollBehavior _configuration;
+  ScrollController? _verticalFallbackController;
+  ScrollController? _horizontalFallbackController;
+  final GlobalKey<ScrollableState> _verticalOuterScrollableKey = GlobalKey<ScrollableState>();
+  final GlobalKey<ScrollableState> _horizontalInnerScrollableKey = GlobalKey<ScrollableState>();
```

Comment:
The global keys are a little sad... This is to implement TwoDimensionalScrollable.of(context).verticalScrollable? Could we just redirect people to Scrollabe.of with axis instead of providing TwoDimensionalScrollable.of? Or we could implement TwoDimensionalScrollable.virticalScrollableOf and have that internally use Scrollabe.of.

Is there any other reason for obtaining a TwoDimensionalScrollableState? Maybe it could even be private?

Reply:
I wanted to still provide this because it is really really clear when folks use it what they are getting.

Using Scrollable.of/maybeOf works, but if I have learned anything about how folks compose scrolling, they will nest.
I feel that in the case where users are nesting a bunch of scrolling widgets together, it can be more difficult to reason/figure out how to achieve what they want with Scrollable and this is a really direct way for them to access these if they want to use them.